### PR TITLE
Add auth spans and remove deduplication code for scopes

### PR DESCRIPTION
**Pull request by jalevin**
**What is this feature?**
This PR adds more spans for timing in accesscontrol. We used the timing insights to investigate some performance bottlenecks in access control phase of requests and found the deduplication of permissions was the culprit. After consulting with @grafana/identity-access-team we decided to remove this.

### CODE REVIEW

File: pkg/services/accesscontrol/acimpl/service.go [hunk 1]
```diff
@@ -437,6 +459,9 @@ func GetActionFilter(options accesscontrol.SearchOptions) func(action string) bo
 
 // SearchUsersPermissions returns all users' permissions filtered by action prefixes
 func (s *Service) SearchUsersPermissions(ctx context.Context, usr identity.Requester, options accesscontrol.SearchOptions) (map[int64][]accesscontrol.Permission, error) {
+	ctx, span := s.tracer.Start(ctx, "authz.SearchUserPermissions")
```
Comment (IevaVasiljeva) [inline pkg/services/accesscontrol/acimpl/service.go:462]:
```suggestion
	ctx, span := s.tracer.Start(ctx, "authz.SearchUsersPermissions")
```

**Review by IevaVasiljeva — APPROVED** [review]
Oh, interesting. Deduplicating permissions is actually something that I added relatively recently (in [this PR](https://github.com/grafana/grafana/pull/87967/files#diff-3f46355ace70434e1c19bdc6f340aa299f8324a0aa87870a3aa1384c6c919d92R235)). My thinking was that deduplicating the scopes once should make any future permission lookups from the cached permissions faster. But I didn't test this change at scale, so if it's too CPU heavy we can undo it. It would also be interesting to see if deduplication does actually lead to any performance benefits - no point doing it if it doesn't.

**Comment by hairyhenderson** [comment]
> Deduplicating permissions is actually something that I added relatively recently

There's a bunch of things going on in that PR and I haven't read very closely so apologies if it's obvious, but: is the `GroupScopesByAction` change such that it can be backed out with little impact?

Also, I'm not convinced that the deduplication _itself_ is to blame for the poor performance in the investigations that @jalevin and I have been doing, but rather the allocations happening in that loop. It's possible that we might not be able to avoid it without some fairly big changes 🤔

**Comment by jalevin** [comment]
> Oh, interesting. Deduplicating permissions is actually something that I added relatively recently (in [this PR](https://github.com/grafana/grafana/pull/87967/files#diff-3f46355ace70434e1c19bdc6f340aa299f8324a0aa87870a3aa1384c6c919d92R235)). My thinking was that deduplicating the scopes once should make any future permission lookups from the cached permissions faster. But I didn't test this change at scale, so if it's too CPU heavy we can undo it. It would also be interesting to see if deduplication does actually lead to any performance benefits - no point doing it if it doesn't.

We added some span attributes to look at reductions. In a configuration with default permissions, 20k users, and 1m dashboards we generate 5095220 permissions. After deduplication we get 5095155 which removed 65 duplicates over 115 actions.

I ran some benchmarks with and without deduplication:

image

I think part of the problem is that we're not actually caching the result of the deduplication so this is happening on every request AFTER gathering permissions from the cached sources. 

Timing of web requests:

deduplicated:
image

non-deduplicated:
image

I'm not entirely sure what configuration could lead to duplicate permissions, however, I'd recommend we either:

1. disable it on permission counts > 100k
2. cache the end result of deduplicated and pull from that cache
3. remove altogether

**Comment by jalevin** [comment]
Consensus reached on option 3, remove deduplication code.

**Pull request merged**